pkg/engine/executor/dataobjscan_predicate.go

```diff
@@ -347,3 +356,132 @@ func mapMetadataPredicate(expr physical.Expression) (logs.RowPredicate, error) {
 		return nil, fmt.Errorf("unsupported expression type (%T) for metadata predicate, expected BinaryExpr or UnaryExpr", expr)
 	}
 }
+
+func mapMessagePredicate(expr physical.Expression) (logs.RowPredicate, error) {
+	switch e := expr.(type) {
+	case *physical.BinaryExpr:
+		switch e.Op {
+
+		case types.BinaryOpMatchSubstr, types.BinaryOpNotMatchSubstr, types.BinaryOpMatchRe, types.BinaryOpNotMatchRe:
+			return match(e)
+
+		case types.BinaryOpMatchPattern, types.BinaryOpNotMatchPattern:
+			return nil, fmt.Errorf("unsupported binary operator (%s) for string predicate", e.Op)
+
+		case types.BinaryOpAnd:
```

Comment:
nit: we would not hit AND and OR cases since we only call `mapMessagePredicate` if the Left is a column expression. https://github.com/grafana/loki/blob/7d90021bfd6b62378facf60296a58f6fe146aa69/pkg/engine/executor/dataobjscan_predicate.go#L68-L75